Well-formedness checking of SSA equation [depends on #3480]

xbauch opened the pull request:
This is an attempt at introducing well-formedness checking into SSA equation. It follows the process of #3123 by implementing the `validate` method in `SSA_stept` and extending `ssa_exprt` with `check/validate` functionality. Unit tests will be included if the attempt is deemed worthy of pursuing.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

danpoe commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -536,6 +536,11 @@ int cbmc_parse_optionst::doit()
+  if(cmdline.isset("validate-ssa-equation"))
+  {
+    options.set_option("validate-ssa-equation", true);
+  }
```
This should best be put in `cbmc_parse_optionst::get_command_line_options()`

xbauch replied:
Done.

danpoe commented on src/util/ssa_expr.h:
```diff
@@ -139,6 +139,21 @@ class ssa_exprt:public symbol_exprt
+    DATA_CHECK(!has_operands(), "SSA expression do not have operands");
+    DATA_CHECK(id() == ID_symbol, "SSA expression symbols are symbols");
+    DATA_CHECK(get_bool(ID_C_SSA_symbol), "wrong SSA expression ID");
+  }
```
Btw, these methods have now become static methods that take the parent type in the recent changes to PR #3123. So that would then become `static void check(const exprt &expr, const validation_modet vm)`.

We should also add a case for `SSA_exprt` to `call_on_expr()` in `validate_expressions.cpp`.

xbauch replied:
Done.

danpoe commented on src/util/ssa_expr.h:
```diff
@@ -139,6 +139,21 @@ class ssa_exprt:public symbol_exprt
+
+  void check(const validation_modet vm = validation_modet::INVARIANT) const
+  {
+    DATA_CHECK(!has_operands(), "SSA expression do not have operands");
```
Nit: According to the coding standard (`cbmc/CODING_STANDARD.md`), if possible the messages in invariants should be "should" statements that describe what should hold, e.g. "SSA expression should have operands".

danpoe reviewed (commented):
Looks good so far, a few minor comments

peterschrammel commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -125,8 +125,29 @@ int goto_instrument_parse_optionst::doit()
     get_goto_program();
 
+    {
+      const bool b = cmdline.isset("validate-goto-binary");
```
What does `b` stand for?

peterschrammel commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1572,6 +1593,12 @@ void goto_instrument_parse_optionst::help()
     " --show-local-safe-pointers   show pointer expressions that are trivially dominated by a not-null check\n" // NOLINT(*)
     " --show-safe-dereferences     show pointer expressions that are trivially dominated by a not-null check\n" // NOLINT(*)
     "                              *and* used as a dereference operand\n" // NOLINT(*)
+    // NOLINTNEXTLINE(whitespace/line_length)
```
I suggest to provide `HELP_VALIDATION` and `OPT_VALIDATION` to avoid code duplication.

peterschrammel commented on src/util/validate.h:
```diff
@@ -0,0 +1,108 @@
+  }
+};
+
+#endif /* CPROVER_UTIL_VALIDATE_H */
```
// preferred

peterschrammel commented on src/util/validate_code.h:
```diff
@@ -0,0 +1,28 @@
+  const namespacet &ns,
+  const validation_modet vm);
+
+#endif /* CPROVER_UTIL_VALIDATE_CODE_H */
```
// preferred

peterschrammel commented on src/util/validate_expressions.h:
```diff
@@ -0,0 +1,28 @@
+  const namespacet &ns,
+  const validation_modet vm);
+
+#endif /* CPROVER_UTIL_VALIDATE_EXPRESSIONS_H */
```
// preferred

peterschrammel commented on src/util/validate_types.h:
```diff
@@ -0,0 +1,28 @@
+  const namespacet &ns,
+  const validation_modet vm);
+
+#endif /* CPROVER_UTIL_VALIDATE_TYPES_H */
```
// preferred

peterschrammel commented on src/util/validate_types.h:
```diff
@@ -0,0 +1,28 @@
+class namespacet;
+enum class validation_modet;
+
+void check_type_pick(const typet &type, const validation_modet vm);
```
Parameter names could be omitted in the declarations. It's obvious what they are.

martin-cs commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -530,6 +530,17 @@ int cbmc_parse_optionst::doit()
+  {
+    const namespacet ns(goto_model.symbol_table);
+    goto_model.validate(ns, validation_modet::INVARIANT);
+  }
```
Is this from a previous PR as it doesn't match the description of the PR?

martin-cs commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -406,6 +406,12 @@ int goto_analyzer_parse_optionst::doit()
     return CPROVER_EXIT_INTERNAL_ERROR;
 
+  if(cmdline.isset("validate-goto-model"))
+  {
```
So, I'm opinionated but ... why add this option to every program?  Why not just add it to goto-instrument?  If it's the same check (and eventually everyone will do it on program load) why do we need the extra option everywhere?

xbauch commented:
Rebased onto #3123 branch to follow the latest version. Only the last commit pertain to SSA checking. Once #3123 is merged into develop, this will be rebased onto that and hopefully cleaner.

tautschnig commented on src/cbmc/bmc.cpp:
```diff
@@ -638,6 +638,10 @@ void bmct::perform_symbolic_execution(
+  if(options.get_bool_option("validate-ssa-equation"))
+  {
+    symex.validate(ns, validation_modet::INVARIANT);
+  }
```
Nit picking: for readability, it would really help to separate the actual work from any checks by, e.g., a blank line.

xbauch replied:
Done.

tautschnig commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -974,3 +974,57 @@ void symex_target_equationt::SSA_stept::output(std::ostream &out) const
+    validate_full_expr(ssa_full_lhs, ns, vm);
+    validate_full_expr(original_full_lhs, ns, vm);
+    validate_full_expr(ssa_rhs, ns, vm);
+    base_type_eq(ssa_lhs.get_original_expr(), ssa_rhs, ns);
```
This has no effect?

xbauch replied:
Oh, my bad: this was intended to check that `lhs` and `rhs` have the same type. Should be fixed now.

tautschnig commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -974,3 +974,57 @@ void symex_target_equationt::SSA_stept::output(std::ostream &out) const
+    const symbolt *symbol;
+    DATA_CHECK(
+      called_function.empty() || !ns.lookup(called_function, symbol),
+      "unknown function identifier " + id2string(called_function));
```
Just making that remark here: could someone please put together a PR that makes the `vm` no longer implicit in here?

xbauch replied:
I have asked @danpoe and we'll have a look at it.

chrisr-diffblue replied:
@danpoe Has just put up #3480 which makes the change to DATA_CHECK macro. Once that PR is merged we should rebase this PR and make the corresponding usage changes here.

tautschnig commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -974,3 +974,57 @@ void symex_target_equationt::SSA_stept::output(std::ostream &out) const
+    validate_full_expr(ssa_lhs, ns, vm);
+    validate_full_expr(ssa_full_lhs, ns, vm);
+    validate_full_expr(original_full_lhs, ns, vm);
+    validate_full_expr(ssa_rhs, ns, vm);
```
We also need to validate that the lhs and rhs of an assignment have the same type.

xbauch replied:
Done.

tautschnig reviewed (commented):
Some comments below, but the biggest item that I'd like to see added/changed is the current `check_renaming` of goto_symex_state.cpp to be moved into the validation framework and thus only done when validation is enabled.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ea8c694471c5e0d648d44780ec2743af897f0c6d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91355754

xbauch commented:
@tautschnig regarding the `check_renaming`: I've tried moving this check into the validation framework. I ended up adding a `validation_modet` variable to both `goto_symext` and `goto_symex_statet` and an `IGNORE` validation mode. Have a look if this is what you had in mind.

tautschnig commented on src/goto-symex/goto_symex.h:
```diff
@@ -76,7 +76,8 @@ class goto_symext
       _total_vccs(std::numeric_limits<unsigned>::max()),
-      _remaining_vccs(std::numeric_limits<unsigned>::max())
+      _remaining_vccs(std::numeric_limits<unsigned>::max()),
+      global_validation_mode(validation_modet::IGNORE)
```
Wouldn't it be possible to look at `options` instead and store that in a Boolean, as is done for several other entries in `options` above?

xbauch replied:
Well I can use `options` in `goto_symext` but not in `statet` (unless I pass it there somehow).

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -241,9 +217,10 @@ void goto_symex_statet::assignment(
   lhs.update_type();
-  assert_l1_renaming(lhs);
+  const validation_modet vm = global_validation_mode;
+  DATA_CHECK(!check_renaming_l1(lhs), "lhs renaming failed on l1");
```
What I'd do here is something along the lines of `!validation_enabled || !check_renaming_l1(lhs)` to avoid changing the `validation_modet` enum.

xbauch replied:
I reverted the new validation mode. I'm not sure if you're suggesting not to use `DATA_CHECK` at all and just call the regular invariants, or not. Have a look at the new version.

tautschnig commented on src/goto-symex/goto_symex.h:
```diff
@@ -524,7 +524,7 @@ class goto_symext
     target.validate(ns, vm);
   }
-  validation_modet global_validation_mode;
+  bool run_validation_checks;
```
With the exception that I'd place this textually next to the other Boolean variables that store options, yes, this is how I thought it could be done.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -217,16 +215,19 @@ void goto_symex_statet::assignment(
+  if(run_validation_checks)
+  {
+    const validation_modet vm = local_validation_mode;
+    DATA_CHECK(!check_renaming_l1(lhs), "lhs renaming failed on l1");
```
Hopefully that pain will go away soon when `DATA_CHECK` takes three arguments...

tautschnig commented on src/goto-symex/symex_main.cpp:
```diff
@@ -301,7 +301,7 @@ void goto_symext::symex_from_entry_point_of(
   statet state;
 
-  state.global_validation_mode = global_validation_mode;
+  state.run_validation_checks = run_validation_checks;
```
One future day we will actually have a proper constructor I suppose...

tautschnig reviewed (commented):
Many thanks for the update!

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 77868dcb1c6a221efa14e360deb1b4cc75652eaf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91655999

xbauch commented:
@tautschnig I've moved the Boolean flag to the previous ones and squashed the commits.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 242655b646d822d5d59e8473c4a041b91caf02c5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91697460

danpoe committed:
Make validation mode parameter explicit in DATA_CHECK() macros

chrisr-diffblue reviewed (approved):
LGTM, but would be good to rebase on top of #3480 and make the implicit parameter to DATA_CHECK be explicit.

danpoe commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -241,15 +215,19 @@ void goto_symex_statet::assignment(
-  assert_l1_renaming(lhs);
-
-  #if 0
+  const validation_modet local_validation_mode = validation_modet::INVARIANT;
```
If the validation mode is hardcoded to `validation_modet::INVARIANT` we could also use `DATA_INVARIANT()` directly instead. I think when using `DATA_CHECK()` the mode should be truly configurable.

danpoe commented on unit/goto-symex/ssa_equation.cpp:
```diff
@@ -0,0 +1,63 @@
+        {
+          caught = true;
+        }
+        REQUIRE(caught);
```
I think catch supports `REQUIRE_THROWS_*` macros to check that a particular exception is thrown.

Commit:
Well-formedness checking for SSA

Follows the #3123 skeleton with additional checks in SSA_stept. Adds a condition
such that renaming checks are only run with the validate option.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: eb0b86bef576c0a742a58d8a392b4cc17669d336).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93144445
